File: packages/agent-protocol-domain/src/app-game-policy-evaluation.ts
import { AppGameSchemaVersion } from '@ocentra-parent/activity-domain/app-game';
import { ActivityEvidenceRefSchema } from '@ocentra-parent/activity-domain/contracts';
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import { AgentEvent, AgentProtocolDefaults, isAgentProtocolLogText, type AgentEventEnvelope } from './contracts';
import { AgentAppGamePolicyReadinessKind } from './app-game-policy-readiness';

const PolicyEvaluationText = Schema.String.pipe(Schema.minLength(1));
const PolicyEvaluationCount = Schema.Number.pipe(Schema.nonNegative(), Schema.int());

const AgentAppGamePolicyEvaluationRequiredReadinessKindSchema = Schema.Literal(
  AgentAppGamePolicyReadinessKind.PolicyEvidence,
  AgentAppGamePolicyReadinessKind.ApprovalAuthority,
  AgentAppGamePolicyReadinessKind.PlatformAuthority,
  AgentAppGamePolicyReadinessKind.AiClassifierContext
);

export const AgentAppGamePolicyEvaluationKind = {
  TimeLimit: 'timeLimit',
  ApprovalRequest: 'approvalRequest',
  CategoryRiskReview: 'categoryRiskReview',
  BlockLaunch: 'blockLaunch',
} as const;

export const AgentAppGamePolicyEvaluationRequestedAction = {
  TimeLimit: 'time-limit',
  AskParent: 'ask-parent',
  Warn: 'warn',
  BlockLaunch: 'block-launch',
} as const;

export const AgentAppGamePolicyEvaluationPolicyAction = {
  TimeLimit: 'time-limit',
  AskParent: 'ask-parent',
  Warn: 'warn',
  Block: 'block',
} as const;

export const AgentAppGamePolicyEvaluationDecisionState = {
  DryRunReady: 'dry-run-ready',
  ManualRequired: 'manual-required',
} as const;

export const AgentAppGamePolicyEvaluationRejectionReason = {
  None: 'none',
  MissingPolicyEvidence: 'missing-policy-evidence',
  MissingApprovalAuthority: 'missing-approval-authority',
  MissingPlatformAuthority: 'missing-platform-authority',
  MissingClassifierContext: 'missing-classifier-context',
  BlockLaunchManualRequired: 'block-launch-manual-required',
} as const;

export const AgentAppGamePolicyEvaluationRowSchema = withParser(
  Schema.Struct({
    schemaVersion: Schema.Literal(AppGameSchemaVersion),
    evaluationId: PolicyEvaluationText,
    evaluationKind: Schema.Literal(
      AgentAppGamePolicyEvaluationKind.TimeLimit,
      AgentAppGamePolicyEvaluationKind.ApprovalRequest,
      AgentAppGamePolicyEvaluationKind.CategoryRiskReview,
      AgentAppGamePolicyEvaluationKind.BlockLaunch
    ),
    requestedAction: Schema.Literal(
      AgentAppGamePolicyEvaluationRequestedAction.TimeLimit,
      AgentAppGamePolicyEvaluationRequestedAction.AskParent,
      AgentAppGamePolicyEvaluationRequestedAction.Warn,
      AgentAppGamePolicyEvaluationRequestedAction.BlockLaunch
    ),
    policyAction: Schema.Literal(
      AgentAppGamePolicyEvaluationPolicyAction.TimeLimit,
      AgentAppGamePolicyEvaluationPolicyAction.AskParent,
      AgentAppGamePolicyEvaluationPolicyAction.Warn,
      AgentAppGamePolicyEvaluationPolicyAction.Block
    ),
    decisionState: Schema.Literal(
      AgentAppGamePolicyEvaluationDecisionState.DryRunReady,
      AgentAppGamePolicyEvaluationDecisionState.ManualRequired
    ),
    rejectionReason: Schema.Literal(
      AgentAppGamePolicyEvaluationRejectionReason.None,
      AgentAppGamePolicyEvaluationRejectionReason.MissingPolicyEvidence,
      AgentAppGamePolicyEvaluationRejectionReason.MissingApprovalAuthority,
      AgentAppGamePolicyEvaluationRejectionReason.MissingPlatformAuthority,
      AgentAppGamePolicyEvaluationRejectionReason.MissingClassifierContext,
      AgentAppGamePolicyEvaluationRejectionReason.BlockLaunchManualRequired
    ),
    reasonCodes: Schema.Array(PolicyEvaluationText),
    requiredReadinessKinds: Schema.Array(AgentAppGamePolicyEvaluationRequiredReadinessKindSchema),
    evidenceReferenceIds: Schema.Array(PolicyEvaluationText),
    evidence: Schema.Array(ActivityEvidenceRefSchema),
    dryRun: Schema.Literal(true),
    enforcementHandoffState: Schema.Literal('disabled'),
    adapterDispatchState: Schema.Literal('not-dispatched'),
  })
);

export const AgentAppGamePolicyEvaluationReadModelSchema = withParser(
  Schema.Struct({
    schemaVersion: Schema.Literal(AppGameSchemaVersion),
    generatedAt: PolicyEvaluationText,
    custodyLabel: PolicyEvaluationText,
    capabilityStatus: PolicyEvaluationText,
    returned: PolicyEvaluationCount,
    policyEvaluationReady: Schema.Boolean,
    manualReviewRequired: Schema.Boolean,
    dryRun: Schema.Literal(true),
    enforcementHandoffState: Schema.Literal('disabled'),
    adapterDispatchClaimed: Schema.Literal(false),
    readinessRowCount: PolicyEvaluationCount,
    evaluatedRowCount: PolicyEvaluationCount,
    evidenceClaimRowCount: PolicyEvaluationCount,
    identityRowCount: PolicyEvaluationCount,
    approvalAuthorityRowCount: PolicyEvaluationCount,
    approvalActionResultRowCount: PolicyEvaluationCount,
    platformAuthorityRowCount: PolicyEvaluationCount,
    aiClassifierResultRowCount: PolicyEvaluationCount,
    rows: Schema.Array(AgentAppGamePolicyEvaluationRowSchema),
  })
);

export type AgentAppGamePolicyEvaluationKind = Infer<typeof AgentAppGamePolicyEvaluationRowSchema>['evaluationKind'];
export type AgentAppGamePolicyEvaluationDecisionState = Infer<
  typeof AgentAppGamePolicyEvaluationRowSchema
>['decisionState'];
export type AgentAppGamePolicyEvaluationRejectionReason = Infer<
  typeof AgentAppGamePolicyEvaluationRowSchema
>['rejectionReason'];
export type AgentAppGamePolicyEvaluationRow = Infer<typeof AgentAppGamePolicyEvaluationRowSchema>;
export type AgentAppGamePolicyEvaluationReadModel = Infer<typeof AgentAppGamePolicyEvaluationReadModelSchema>;

export type AgentAppGamePolicyEvaluationFailureReason =
  | 'wrong-event'
  | 'missing-json-field'
  | 'invalid-json'
  | 'invalid-payload';

export type AgentAppGamePolicyEvaluationResult =
  | {
      readonly ok: true;
      readonly value: AgentAppGamePolicyEvaluationReadModel;
    }
  | {
      readonly ok: false;
      readonly reason: AgentAppGamePolicyEvaluationFailureReason;
    };

export function parseAgentAppGamePolicyEvaluationEvent(event: AgentEventEnvelope): AgentAppGamePolicyEvaluationResult {
  if (event.event !== AgentEvent.ActivityAppGamePolicyEvaluationReadModelReported) {
    return adapterFailure('wrong-event');
  }

  const raw = event.payload[AgentProtocolDefaults.Field.ActivityAppGamePolicyEvaluationReadModel];
  if (!isAgentProtocolLogText(raw)) {
    return adapterFailure('missing-json-field');
  }

  let decoded: unknown;
  try {
    decoded = JSON.parse(raw);
  } catch {
    return adapterFailure('invalid-json');
  }

  const parsed = AgentAppGamePolicyEvaluationReadModelSchema.safeParse(decoded);
  if (!parsed.success || parsed.data === undefined) {
    return adapterFailure('invalid-payload');
  }

  return {
    ok: true,
    value: parsed.data,
  };
}

function adapterFailure(reason: AgentAppGamePolicyEvaluationFailureReason): AgentAppGamePolicyEvaluationResult {
  return {
    ok: false,
    reason,
  };
}

File: crates/agent-protocol/src/app_game_policy_evaluation.rs
use serde::{Deserialize, Serialize};

use crate::ActivityEvidenceRef;

pub const APP_GAME_POLICY_EVALUATION_CUSTODY_CHILD_DEVICE_QUERY_STORE: &str =
    "child-device-query-store";
pub const APP_GAME_POLICY_EVALUATION_STATUS_READY: &str = "policy-evaluation-ready";
pub const APP_GAME_POLICY_EVALUATION_STATUS_MANUAL_REQUIRED: &str =
    "policy-evaluation-manual-required";
pub const APP_GAME_POLICY_EVALUATION_STATUS_NO_ROWS: &str = "policy-evaluation-no-rows";
pub const APP_GAME_POLICY_EVALUATION_KIND_TIME_LIMIT: &str = "timeLimit";
pub const APP_GAME_POLICY_EVALUATION_KIND_APPROVAL_REQUEST: &str = "approvalRequest";
pub const APP_GAME_POLICY_EVALUATION_KIND_CATEGORY_RISK_REVIEW: &str = "categoryRiskReview";
pub const APP_GAME_POLICY_EVALUATION_KIND_BLOCK_LAUNCH: &str = "blockLaunch";
pub const APP_GAME_POLICY_EVALUATION_REQUESTED_ACTION_TIME_LIMIT: &str = "time-limit";
pub const APP_GAME_POLICY_EVALUATION_REQUESTED_ACTION_ASK_PARENT: &str = "ask-parent";
pub const APP_GAME_POLICY_EVALUATION_REQUESTED_ACTION_WARN: &str = "warn";
pub const APP_GAME_POLICY_EVALUATION_REQUESTED_ACTION_BLOCK_LAUNCH: &str = "block-launch";
pub const APP_GAME_POLICY_EVALUATION_POLICY_ACTION_TIME_LIMIT: &str = "time-limit";
pub const APP_GAME_POLICY_EVALUATION_POLICY_ACTION_ASK_PARENT: &str = "ask-parent";
pub const APP_GAME_POLICY_EVALUATION_POLICY_ACTION_WARN: &str = "warn";
pub const APP_GAME_POLICY_EVALUATION_POLICY_ACTION_BLOCK: &str = "block";
pub const APP_GAME_POLICY_EVALUATION_DECISION_DRY_RUN_READY: &str = "dry-run-ready";
pub const APP_GAME_POLICY_EVALUATION_DECISION_MANUAL_REQUIRED: &str = "manual-required";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_NONE: &str = "none";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_MISSING_POLICY_EVIDENCE: &str =
    "missing-policy-evidence";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_MISSING_APPROVAL_AUTHORITY: &str =
    "missing-approval-authority";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_MISSING_PLATFORM_AUTHORITY: &str =
    "missing-platform-authority";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_MISSING_CLASSIFIER_CONTEXT: &str =
    "missing-classifier-context";
pub const APP_GAME_POLICY_EVALUATION_REJECTION_BLOCK_LAUNCH_MANUAL_REQUIRED: &str =
    "block-launch-manual-required";
pub const APP_GAME_POLICY_EVALUATION_REASON_READY: &str = "app-game-policy-readiness-ready";
pub const APP_GAME_POLICY_EVALUATION_REASON_ADAPTER_DISPATCH_DISABLED: &str =
    "adapter-dispatch-disabled";
pub const APP_GAME_POLICY_EVALUATION_REASON_MANUAL_REQUIRED: &str =
    "app-game-policy-manual-required";
pub const APP_GAME_POLICY_EVALUATION_HANDOFF_DISABLED: &str = "disabled";
pub const APP_GAME_POLICY_EVALUATION_ADAPTER_NOT_DISPATCHED: &str = "not-dispatched";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGamePolicyEvaluationRow {
    pub schema_version: u16,
    pub evaluation_id: String,
    pub evaluation_kind: String,
    pub requested_action: String,
    pub policy_action: String,
    pub decision_state: String,
    pub rejection_reason: String,
    pub reason_codes: Vec<String>,
    pub required_readiness_kinds: Vec<String>,
    pub evidence_reference_ids: Vec<String>,
    pub evidence: Vec<ActivityEvidenceRef>,
    pub dry_run: bool,
    pub enforcement_handoff_state: String,
    pub adapter_dispatch_state: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGamePolicyEvaluationReadModel {
    pub schema_version: u16,
    pub generated_at: String,
    pub custody_label: String,
    pub capability_status: String,
    pub returned: u64,
    pub policy_evaluation_ready: bool,
    pub manual_review_required: bool,
    pub dry_run: bool,
    pub enforcement_handoff_state: String,
    pub adapter_dispatch_claimed: bool,
    pub readiness_row_count: u64,
    pub evaluated_row_count: u64,
    pub evidence_claim_row_count: u64,
    pub identity_row_count: u64,
    pub approval_authority_row_count: u64,
    pub approval_action_result_row_count: u64,
    pub platform_authority_row_count: u64,
    pub ai_classifier_result_row_count: u64,
    pub rows: Vec<AppGamePolicyEvaluationRow>,
}
